### ISSUE TITLE
Refactoring lua files

### ISSUE THREAD

**Issue by songgao** [virgo-agent-toolkit/rackspace-monitoring-agent#443]
TL;DR: division in `virgo` has some design issues to address. I'm gonna proceed with a little hacky way for now, but we need to think about the design for long-term run.

Currently all lua scripts, monitoring-related or not, are in `virgo` instead `virgo-base`. This is fine for people who just wanna deploy their own monitoring system, but not good for people who wanna develop their own agent based on `virgo-base`, e.g., a cloudkeep(https://github.com/cloudkeep) agent that uses FUSE to provide keys to applications. The lua files in `virgo` need to be divided into monitoring-related part (which stays in `virgo`) and general part (which should be moved into `virgo-base`.

Here's a diagram I drew while reading:
![old](https://f.cloud.github.com/assets/255797/858823/531e2700-f56f-11e2-84df-c0f7540f2cdd.png)

Red nodes are the ones with monitoring logic. Some of them are easy to deal with, like `/check/*.lua`, `/schedule/scheduler.lua`, since they are independent modules that can be simply isolated from the core part. The challenging part is the _MonitoringAgent -> ConnectionStream -> AgentClient -> AgentProtocolConnection_ path. Each of them more of less has something to do with monitoring.

@robert-chiniquy and I thought about providing a base type for each of them, which has core logic such as handshake.hello (authentication) and heartbeat, and use dependency injection to determine what type to use. When users need to extend something, they simply sub-class the base type and inject the type to proper place.

However, the idea above has issues since there are too many levels. Suppose we need to add a new RPC call, `helloworld`, which uses a new message type `helloworld.param`. The message type is not a big issue. But we also need to add the handler for `helloworld` RPC method, which is defined in `/protocol/connection.lua`. However, it's in the bottom level. In order to support that, the type needs to be specified somehow from top down.

For now, so we are gonna add an argument to constructors, which has types that needs to be overwritten. Then when instantiating objects, if a type is defined for that object, use that type; other wise, use the default one. But this is not a good design in long-term. We probably still need to refactor those scripts in the future for a cleaner design.
